### PR TITLE
[Language Server] Type and reference accumulaton

### PR DESCRIPTION
**Pull request by ysndr**

### CODE REVIEW

**Comment by ysndr** [comment]
- [x] hover (in-line types too maybe)
- [x] references and usages
- [x] record completion

Next: figuring out the front end communication

**Comment by ysndr** [comment]
Go to **references** is implemented. But bot having the identifier in the AST means we cannot actually point to anything that declares a referenced variable.

File: src/typecheck/mod.rs [hunk 1]
```diff
@@ -438,22 +446,39 @@ pub struct State<'a> {
...
 /// [`type_check_`](fn.type_check_.html) with a fresh unification variable as goal.
-pub fn type_check(
+pub fn type_check<'a, L>(
     t: &RichTerm,
     global_eval_env: &eval::Environment,
-    resolver: &dyn ImportResolver,
-) -> Result<Types, TypecheckError> {
-    let mut state = State {
+    resolver: &'a impl ImportResolver,
+    linearizer: impl Linearizer<L, UnifTable>,
```
[the remark this answers is not in the document]
Comment (yannham) [reply src/typecheck/mod.rs:454]: Oops, didn't see it's a wrapper. It wasn't actually  needed. I'm gonna fix this myself before merging.

File: src/typecheck/mod.rs [hunk 2]
```diff
@@ -522,6 +524,8 @@ pub fn type_check_in_env(
 ///
 /// - `state`: the unification state (see [`State`](struct.State.html)).
 /// - `env`: the typing environment, mapping free variable to types.
+/// - `lin`: The current building linearization of building state `S`
+/// - `linearizer`: A linearizer that can madify the linearization
```
Comment (yannham) [inline src/typecheck/mod.rs:528]:
```suggestion
/// - `linearizer`: A linearizer that can modify the linearization
```

**Review by yannham — APPROVED** [review]

**Pull request merged**